src/iris/api.py

```diff
@@ -3694,6 +3696,138 @@ def on_get(self, req, resp, app_name):
+                          VALUES (%s,
+                                  %s,
+                                  (SELECT `id` FROM `target` WHERE `name` = %s
+                                     AND `type_id` = (SELECT `id` FROM `target_type` WHERE `name` = 'user')))''',
+                       (device_uuid, os, user))
+        conn.commit()
+        cursor.close()
+        conn.close()
+
+        # disconnect from server
+        resp.status = falcon.HTTP_201
+
+
+# FIXME: temporary check to limit allowed users to a small group
+def check_allowed(req, resp, resource, params):
+    if req.context['username'] not in resource.allowed_users:
+        raise falcon.HTTPForbidden()
+
+
+@falcon.before(check_allowed)
+class QRKey(object):
+    allow_read_no_auth = False
+
+    def __init__(self, mobile_url, users):
+        self.mobile_url = mobile_url
+        self.allowed_users = users
+
+    def png_to_datauri(self, png):
+        return 'data:image/png;base64,%s' % png
+
+    def generate_secret(self, user):
+        # Generate crypto-secure 256 bit secret
+        random_bytes = os.urandom(32)
+        secret = b64encode(random_bytes).decode('utf-8')
+
+        data = {'secret': secret,
+                'api_url': self.mobile_url}
+        code = pyqrcode.create(ujson.dumps(data))
+        image_as_str = code.png_as_base64_str(scale=5)
+
+        # Open database connection
+        conn = db.engine.raw_connection()
+        cursor = conn.cursor()
+
+        try:
+            cursor.execute(
+                '''UPDATE `user` SET `mobile_key` = %s
+                   WHERE `target_id` = (SELECT `id` FROM `target` WHERE `name` = %s
+                     AND `type_id` = (SELECT `id` FROM `target_type` WHERE `name` = 'user'))''',
+                (secret, user))
+            if cursor.rowcount == 0:
+                raise HTTPNotFound()
+            # Commit your changes in the database
+            conn.commit()
+        except falcon.HTTPError:
+            raise
+        except:
+            logger.exception('Failed to generate key for user %s', user)
+            raise falcon.HTTPError('Error generating key')
+        finally:
+            cursor.close()
+            conn.close()
+
+        return self.png_to_datauri(image_as_str)
+
+    def on_get(self, req, resp, user):
+        conn = db.engine.raw_connection()
+        cursor = conn.cursor()
+
+        cursor.execute(
+            '''SELECT `mobile_key` FROM `user`
+               WHERE `target_id` = (SELECT `id` FROM `target` WHERE `name` = %s
+                 AND `type_id` = (SELECT `id` FROM `target_type` WHERE `name` = 'user'))''',
+            user)
+        secret = cursor.fetchone()
+        if secret is None:
+            cursor.close()
+            conn.close()
+            raise HTTPNotFound()
+        else:
+            secret = secret[0]
+        data = {'secret': secret,
+                'api_url': self.mobile_url}
+        cursor.close()
```

Comment:
You can move the cursor and connection closing calls to right after cursor.fetchone() as it doesn't look like they're used below, that way you can avoid closing them in two places.